src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/DefaultRazorIntermediateNodeLoweringPhase.cs

```diff
@@ -2225,6 +2128,33 @@ private void Combine(HtmlContentIntermediateNode node, SyntaxNode item)
         }
     }
 
+    private ref struct DirectiveAttributeName(ReadOnlySpan<char> original)
+    {
+        // Directive attributes should start with '@' unless the descriptors are misconfigured.
+        // In that case, we would have already logged an error.
+        public readonly ReadOnlySpan<char> Span = original.StartsWith('@') ? original[1..] : original;
+
+        private bool? _hasParameter;
+
+        public string Text => field ??= Span.ToString();
+
+        public bool HasParameter => _hasParameter ??= Span.IndexOf(':') >= 0;
+
+        public string TextWithoutParameter
+        {
+            get
+            {
+                return field ??= GetWithoutParameter(Span);
+
+                static string GetWithoutParameter(ReadOnlySpan<char> span)
+                {
+                    var index = span.IndexOf(':');
+                    return index >= 0 ? span[..index].ToString() : span.ToString();
```

Comment:
>span.ToString() [](http://example.com/codeflow?start=67&length=15)

Not sure how common this is, but couldn't this use the Text property to (likely) avoid this allocation?